src/goto-instrument/contracts/dynamic-frames/dfcc_instrument_loop.cpp

```diff
@@ -0,0 +1,717 @@
+  spec_functions.generate_havoc_instructions(
+    function_id,
+    language_mode,
+    symbol_table.get_writeable_ref(function_id).module,
+    assigns_instrs,
+    addr_of_loop_write_set,
+    havoc_instrs,
+    nof_targets);
+  spec_functions.to_spec_assigns_instructions(
+    addr_of_loop_write_set, language_mode, assigns_instrs, nof_targets);
+  write_set_populate_instrs.copy_from(assigns_instrs);
+
+  // replace bound variables by fresh instances in quantified formulas
+  exprt invariant = loop.invariant;
+  if(has_subexpr(invariant, ID_exists) || has_subexpr(invariant, ID_forall))
+    add_quantified_variable(symbol_table, invariant, language_mode);
+
+  // ---------- Add instrumented instructions ----------
+  const symbol_exprt &loop_write_set = loop.write_set_var;
+  goto_programt::targett loop_latch =
+    loop.find_latch(goto_function.body).value();
+
+  const auto &history_var_map = add_prehead_instructions(
+    loop_id,
+    goto_function,
+    symbol_table,
+    head,
+    loop_latch,
+    write_set_populate_instrs,
+    invariant,
+    assigns,
+    loop_write_set,
+    addr_of_loop_write_set,
+    entered_loop,
+    initial_invariant,
+    in_base_case,
+    language_mode);
+
+  const auto step_target = add_step_instructions(
+    loop_id,
+    cbmc_loop_id,
+    function_id,
+    goto_function,
+    symbol_table,
+    head,
+    loop_latch,
+    havoc_instrs,
+    invariant,
+    decreases_clauses,
+    addr_of_loop_write_set,
+    outer_write_set,
+    initial_invariant,
+    in_base_case,
+    old_decreases_vars,
+    language_mode);
+
+  add_body_instructions(
+    loop_id,
+    cbmc_loop_id,
+    goto_function,
+    symbol_table,
+    head,
+    loop_latch,
+    invariant,
+    decreases_clauses,
+    entered_loop,
+    in_base_case,
+    old_decreases_vars,
+    new_decreases_vars,
+    step_target,
+    language_mode);
+
+  add_exit_instructions(
+    loop_id,
+    cbmc_loop_id,
+    goto_function,
+    head,
+    decreases_clauses,
+    loop_write_set,
+    addr_of_loop_write_set,
+    history_var_map,
+    entered_loop,
+    initial_invariant,
+    in_base_case,
+    old_decreases_vars,
+    new_decreases_vars);
+
+  goto_function.body.update();
+}
+
+std::map<exprt, exprt> dfcc_instrument_loopt::add_prehead_instructions(
+  const std::size_t loop_id,
+  goto_functionst::goto_functiont &goto_function,
+  symbol_tablet &symbol_table,
+  goto_programt::targett loop_head,
+  goto_programt::targett loop_latch,
+  goto_programt &assigns_instrs,
+  exprt &invariant,
+  const exprt::operandst &assigns,
+  const symbol_exprt &loop_write_set,
+  const symbol_exprt &addr_of_loop_write_set,
+  const symbol_exprt &entered_loop,
+  const symbol_exprt &initial_invariant,
+  const symbol_exprt &in_base_case,
+  const irep_idt &language_mode)
+{
+  auto loop_head_location(loop_head->source_location());
+  dfcc_remove_loop_tags(loop_head_location);
+  goto_convertt converter(symbol_table, message_handler);
```

Comment:
Please move this to the (only) block that actually uses it.

Reply:
currently it is needed for the assert(invariant), assume(invariant) and decreases clause expressions

Reply:
And also `initial_invariant = invariant
`